src/puya/awst_build/eb/log.py

```diff
@@ -0,0 +1,123 @@
+        try:
+            sep_index = arg_names.index("sep")
+        except ValueError:
+            sep: Expression = BytesConstant(value=b"", source_location=location)
+        else:
+            sep = expect_operand_wtype(args_.pop(sep_index), wtypes.bytes_wtype)
+
+        log_value: Expression | None = None
+        for arg in args_:
+            match arg:
+                case ExpressionBuilder(value_type=wtypes.uint64_wtype):
+                    bytes_expr = itob(arg.rvalue(), arg.source_location)
+                case Literal(value=int(int_literal)):
+                    bytes_expr = itob(
+                        UInt64Constant(value=int_literal, source_location=arg.source_location),
+                        arg.source_location,
+                    )
+                case ExpressionBuilder(value_type=wtypes.bytes_wtype):
+                    bytes_expr = arg.rvalue()
+                case ExpressionBuilder(value_type=wtypes.biguint_wtype):
+                    bytes_expr = ReinterpretCast(
+                        expr=arg.rvalue(),
+                        wtype=wtypes.bytes_wtype,
+                        source_location=arg.source_location,
+                    )
+                case Literal(value=bytes(bytes_literal)):
+                    bytes_expr = BytesConstant(
+                        value=bytes_literal, source_location=arg.source_location
+                    )
+                case Literal(value=str(str_literal)):
+                    bytes_expr = BytesConstant(
+                        value=str_literal.encode("utf8"), source_location=arg.source_location
+                    )
+                case ExpressionBuilder() as eb:
+                    raise CodeError(
+                        f"Unexpected argument type: {eb.value_type}", arg.source_location
+                    )
+                case Literal() as lit:
+                    raise CodeError(
+                        f"Unexpected argument type: {type(lit.value).__name__}",
+                        arg.source_location,
+                    )
+                case _:
+                    raise CodeError("Unexpected argument", arg.source_location)
+            if log_value is None:
+                log_value = bytes_expr
+            elif sep:
+                log_value = concat(
+                    concat(log_value, sep, arg.source_location),
+                    bytes_expr,
+                    arg.source_location,
+                )
+            else:
+                log_value = concat(log_value, bytes_expr, arg.source_location)
+        if log_value is None:
+            log_value = BytesConstant(value=b"", source_location=location)
+        return var_expression(
+            IntrinsicCall(
+                op_code="log",
+                wtype=wtypes.void_wtype,
+                stack_args=[log_value],
+                source_location=location,
+            )
+        )
+
+
+def itob(expr: Expression, loc: SourceLocation) -> Expression:
```

Comment:
these should preferably be underscore prefixed, so we don't get them as suggested imports etc